### ISSUE TITLE
Asynchronous indexing can break large servers

### ISSUE THREAD

**Issue by gamort** [magento/magento2#30012]
Large stores with extremely active product databases can suffer many related problems with indexers not being kept up to date.  Products can disappear from the website, product data can be out of date on the website, excessive cache flushing can  cause huge server load spikes.

This is due to the manner in which triggers are created by the core magento code in https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/lib/internal/Magento/Framework/Mview/View/Subscription.php#L201

INSERT INTO  will insert a new record regardless of whether one exists or not.
REPLACE INTO will as the name says, replace a record - and in this case it will generate a new version_id since that is an auto_increment field.

### Preconditions (*)
1. An extremely large database.  Example 500,000 products
2. An active update schedule

### Steps to reproduce (*)
1.  Create a sample database with 500,000 products
2. Enable Asynchronous indexing
3. Reduce crons to every 10 minutes
4. Run a mass script to change the inventory for the products every 30 seconds.   So within 10 minutes, you will have made 10 million updates

### Expected result (*)
1. When the indexer runs, it will index the latest product information once for each index

### Actual result (*)
It will actually try to index each product 20 times, once for each update

### Additional information to reproduce the issue

1. Change the price of the product with SKU `24-MB01` 20 times via API `rest/all/V1/products/24-MB01`
2. Doing the price change data inserted in the table `catalog_product_price_cl`
3. Adjust the `DEFAULT_BATCH_SIZE` in file `lib/internal/Magento/Framework/Mview/View.php` with 10 from 1000.
4. Enable the DB logs via the `bin/magento dev:query-log:enable` command.
5. grep the text to get the SQL query on table `catalog_product_price_cl`. Please find attached the screenshot of db.log:

<img width="1428" alt="image" src="https://user-images.githubusercontent.com/51681618/201119571-8837ea56-43b4-4330-aae0-879a6134e719.png">

The query always returns `entity_id 1` each time.

In this case, it seems that entity_id 1 will process many times, instead of only 1 time, and should be on the latest one.
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._

**Comment by hostep** [comment]
Issue makes sense I think, removing stale issue label...

@sidolov: would be great if somebody could verify if this is still the case in 2.4-develop (since the referenced code was refactored a couple of months ago) and then assign priorities.

**Comment by hostep** [comment]
Sigh ...

@sidolov, @ihor-sviziev : can we please delete this stalebot crappy thing?
What's the use of people opening issues if nobody looks at them and then they get auto closed? That's just depressing and might scare people away from opening more issues in the future ...

(not saying that this issue is valid, but it feels like it should at least be investigated instead of auto closed)

**Comment by ihor-sviziev** [comment]
@sidolov @sivaschenko @gabrieldagama this should be p1 or p2 priority

**Comment by sidolov** [comment]
@hostep , @ihor-sviziev  we will take a look on it and prioritize in case it' still reproducible

**Comment by engcom-Foxtrot** [comment]
Hello, @gamort. I've checked this issue on the 2.4-develop branch and seems like it's already fixed. Could you please check it on the latest codebase?

**Comment by onlinebizsoft** [comment]
I think we have same problem with 2.4.2, keep getting lock in different user cases

An error occurred during creating the deployment package, error message: Unable to execute command "php7.4 /tmp/builds/d-3e55edc1/bin/magento setup:upgrade 2>&1" - Running schema recurring...SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: UPDATE `cron_schedule` SET `status` = ?, `messages` = ? WHERE (status = 'running')

**Comment by onlinebizsoft** [comment]
And maybe it is why this extension made ?
https://github.com/magemojo/m2-ce-cron

**Comment by engcom-Hotel** [comment]
Hello @gamort,

Please try to reproduce this issue in the 2.4-develop branch and let us know if you are still able to reproduce the issue.

Thanks

**Comment by fgarnier-hublot** [comment]
Hello @hostep 

I'll share my final conclusion after some test.

I migrated from 2.3.4 to 2.4.3-p1 recentrly.
After a full product and stock import, I have some very long scheduled index and even some are stuck forever.
It creates 800k line with only 30k distinct for "catalogrule_product_cl" as an Example.

(I have 0 rule in place, but the indexer takes more than 3 days, never finished... this is another issue, a foreach product than a foreach rule)

But for me, by reading the source code,

There is an issue when the mview is being updated if the same entity_id is present in two different chunk of the batch size.
On 2.3.4, they first retrieve the max 100k versions, make them unique and batch on them

**2.3.4**
https://github.com/magento/magento2/blob/5f3b86ab4bd3e3b94e65429fed33f748d29c1bbe/lib/internal/Magento/Framework/Mview/View.php#L292-L311

https://github.com/magento/magento2/blob/5f3b86ab4bd3e3b94e65429fed33f748d29c1bbe/lib/internal/Magento/Framework/Mview/View/Changelog.php#L145-L166

**3.4.3-p1**

https://github.com/magento/magento2/blob/8afdb9858d238392ecb5dbfe556068ec1af137bc/lib/internal/Magento/Framework/Mview/View.php#L299-L315

https://github.com/magento/magento2/blob/8afdb9858d238392ecb5dbfe556068ec1af137bc/lib/internal/Magento/Framework/Mview/View/ChangeLogBatchWalker.php#L35-L58

Franck.

**Comment by ihor-sviziev** [comment]
Hi @fgarnier-hublot,
Could you report it as a separate issue?
Thank you!

**Comment by xhoaraujx** [comment]
Hi, 
sorry to respond on a closed issue, but I think this is still relevant and I didn't found any other related issue.

The problem clearly comes from the ChangeLogBatchWalker::walk() function.


It misses a limit offset so it actually loops multiple times on the same values.
Here is a simple example:

Fill your changelog table with 10000 rows of only 3 distinct entity_ids (1, 2, 3)
Default batching is 1000.

walk() function will perform a SELECT DISTINCT on rows 0 to 10000 with LIMIT 1000
> it will return: 1, 2, 3


on second call, it will perform a SELECT DISTINCT on rows 1000 to 10000 with LIMIT 1000
> which also returns 1, 2, 3 ... 

same for 8 more useless calls!


the easy fix would be to add an offset argument to the walk function 
so, on second call, we could perform a SELECT DISTINCT on rows 0 to 10000 with LIMIT 1000 OFFSET 1000
which will return, in our example, no result, and end the walk loops

Hope it helps,
Regards

**Comment by hostep** [comment]
@ihor-sviziev: can this be re-opened and re-validated? Now that there is more info.

**Comment by engcom-Hotel** [comment]
Hello @hostep,

Let me reopen this for further investigation.

Thanks

**Comment by CHallski** [comment]
Having had this crushing our relatively small store (few thousand products, a dozen customer groups, a dozen catalog price rules) by hanging up the process for indexer_update_all_views for hours: I'm trying a different approach.

Rather than building out a massive batch walker update to deal with multiple version_ids per entity_id and get the max for each one, I'm updating the generic table definitions and triggers so that:

- The entity_id is inserted if it isn't present already
- If it already exists the version_id is updated to the newest auto_increment value since all we really care about is that the "latest" version_id in the changelog table for an entity_id is greater than the pointer for that table stored in mview_state.

Yes, there's additional overhead, but it's orders of magnitude less than churning the indexers constantly. This still required manual intervention listed below to deploy, but until the index is added the code continues to function similarly to Magento's current implementation.
- Setting indexers to "update on save"
- Flushing cache
- Deleting the existing _cl tables manually
- Reset the mview_state.version_id to 0 for all indexers
- Setting the indexers back to "update by schedule" to recreate the _cl tables
- Invalidating existing indexes and reindexing to ensure all changes are captured

[fix-nonunique-changelog-indexers.txt](https://github.com/magento/magento2/files/9973639/fix-nonunique-changelog-indexers.txt)

**Comment by bigbangx** [comment]
> Having had this crushing our relatively small store (few thousand products, a dozen customer groups, a dozen catalog price rules) by hanging up the process for indexer_update_all_views for hours: I'm trying a different approach.
> 
> Rather than building out a massive batch walker update to deal with multiple version_ids per entity_id and get the max for each one, I'm updating the generic table definitions and triggers so that:
> 
> * The entity_id is inserted if it isn't present already
> * If it already exists the version_id is updated to the newest auto_increment value since all we really care about is that the "latest" version_id in the changelog table for an entity_id is greater than the pointer for that table stored in mview_state.
> 
> Yes, there's additional overhead, but it's orders of magnitude less than churning the indexers constantly. This still required manual intervention listed below to deploy, but until the index is added the code continues to function similarly to Magento's current implementation.
> 
> * Setting indexers to "update on save"
> * Flushing cache
> * Deleting the existing _cl tables manually
> * Reset the mview_state.version_id to 0 for all indexers
> * Setting the indexers back to "update by schedule" to recreate the _cl tables
> * Invalidating existing indexes and reindexing to ensure all changes are captured
> 
> [fix-nonunique-changelog-indexers.txt](https://github.com/magento/magento2/files/9973639/fix-nonunique-changelog-indexers.txt)

As the REPLACE statement is slower than INSERT, I think that approach could impact triggers performances and slow down large operations.

Also, what about replacement occurring while mview indexation is running ? Some rows that would have been processed during this cron process, would be delayed to next cron, right ?

**Comment by engcom-Hotel** [comment]
Hello @gamort @hostep @bigbangx,

Thanks for your input on this issue. I have some findings and would like to share them here. In order to try to reproduce the issue, I have followed the below steps:

1. Change the price of the product with SKU `24-MB01` 20 times via API `rest/all/V1/products/24-MB01`
2. Doing the price change data inserted in the table `catalog_product_price_cl`
3. Adjust the `DEFAULT_BATCH_SIZE` in file `lib/internal/Magento/Framework/Mview/View.php` with 10 from 1000.
4. Enable the DB logs via the `bin/magento dev:query-log:enable` command.
5. grep the text to get the SQL query on table `catalog_product_price_cl`. Please find attached the screenshot of db.log:

<img width="1428" alt="image" src="https://user-images.githubusercontent.com/51681618/201119571-8837ea56-43b4-4330-aae0-879a6134e719.png">

The query always returns `entity_id 1` each time.

In this case, it seems that entity_id 1 will process many times, instead of only 1 time, and should be on the latest one.

Thanks

**Comment by CHallski** [comment]
> As the REPLACE statement is slower than INSERT, I think that approach could impact triggers performances and slow down large operations.
> 
> Also, what about replacement occurring while mview indexation is running ? Some rows that would have been processed during this cron process, would be delayed to next cron, right ?

@bigbangx You are correct that it does impact trigger performance, but we weighed the impact to doing a catalog-wide update as a couple of seconds increased processing time for our store. Impact to saving product changes and other real-time actions like stock updates was negligible. We had far more problems with the backlog reprocessing changes on the same entities than we are having with the version_id being replaced.

Regarding the version_id updating *during* indexing: the entity_id has already been selected at that point, so it's going to be indexed. The max version_id pointer for that batch should also already be saved, so what should happen is:
- The entity in question will be indexed in the current run
- the mview_state.version_id pointer will be set to the highest one selected for processing from the time the entity_ids for indexing were collected (I would double-check on this one and confirm that they are saving that pointer's value at the time of collection and not re-selecting the highest one after indexing a batch is completed - my understanding of the functions was that they had limits that they were processing set with fromVersion and toVersion parameters)
- The version_id for the entity in question will be incremented past the current pointer
- the next reindex will capture that the entity needs to be reindexed again, and process it during that pass.

This is essentially the same process as now, with the difference that if you have multiple iterations of the same entities being changed by multiple bulk actions, you won't get the **oldest** version ids for processing and work through the same set repeatedly.

The extreme edge case would be if you had so many products being updated that your indexer _never_ caught up to the version_id of a particular entity because there were always more entities with older version_ids filling up your batch size. If you have that problem, it doesn't matter how this portion is coded: your catalog churn is faster than Magento's current indexer model can keep up with.

**Comment by CHallski** [comment]
With all that being said: there are other issues that should be addressed with the current architecture like the sheer number of changelog entries that can be triggered: auto_increment int fields do run out of digits eventually. This was an expeditious band-aid that could be patched in to keep things running while an optimal long-term solution is explored.

**Comment by engcom-Hotel** [comment]
We are confirming this issue as per our findings here: https://github.com/magento/magento2/issues/30012#issuecomment-1310391317

Also updated the main description to reproduce the issue.

Thanks

**Comment by mpchadwick** [comment]
I don't think this was specifically covered in the initial report, but on a related note  we have discovered a major functional problem with the Magento partial indexer (indexer_update_all_views) starting with Magento 2.4.3 including the latest version (2.4.6)

The indexer processes records from the changelog in batches. It gets the IDs to process here:

https://github.com/magento/magento2/blob/2.4.6/lib/internal/Magento/Framework/Mview/View.php#L308

This will generate a query like this

SELECT DISTINCT $changeLogTable.entity_id FROM changeLogTable WHERE (version_id > $fromVersionId) AND (version_id <= $toVersionId) GROUP BY `entity_id` LIMIT $batchSize

This will SELECT the 1000 rows greater than the $fromVersionId all the way to $toVersionId. As a result of the GROUP BY they will also be ordered by the entity_id

So let's say you have a large backlog greater than fromVersionId and have entity_ids starting 1,4,7 etc...

It will process those, even if they are not in the range of $fromVersionId + $batchSize

Next it will increment fromVersionId by batchSize and then run through the loop again

https://github.com/magento/magento2/blob/2.4.6/lib/internal/Magento/Framework/Mview/View.php#L313

This means with a large backlog it is not guaranteed that the entities from the previous batch are actually the ones that get processed.

I'm observing this on a project right now with a very large backlog and what it's doing is just processing the same set of entity_ids over and over for each batch, which are the lowest ones and aren't even the ones that are in the batch it's supposed to be processing. Essentially none of the data is indexed correctly by the partial indexer.

**Comment by sprankhub** [comment]
@mpchadwick, solid analysis as always, thanks! By chance, does this also mean that the partial indexing never finishes or at least runs _very_ long if it processes the same IDs over and over again?

**Comment by mpchadwick** [comment]
@sprankhub - It goes through the backlog based on the batch size. It should take around the same amount of time regardless of the issue I outlined above as the backlog size would be the same. It's just that with the issue I described in each batch it can process the wrong entities, and then move forward to the next batch and the entities that should have been processed from previous batch never get processed.

**Comment by sprankhub** [comment]
Since it is rather hidden in this issue, I would like to mention again that there is a [promising pull request](https://github.com/magento/magento2/pull/36155) already, which might fix the issue.

**Comment by fredden** [comment]
I wonder if the answer here is to throw away the `auto_increment` column from the change-log tables (`*_cl`), and only store the ID of the entity that's changed (if needed, with a `UNIQUE` constraint and use `insert or ignore into ...` in the trigger to avoid duplicates); on the indexer side we can delete values when they're processed. The "how many in change-log table" statistic can then be `select count(*) from table_cl` instead of tracking the last-seen ID and searching for any new values. This also solves the problem of running out of `auto_increment` numbers in the change-log table for a busy catalogue (which doesn't seem to be part of this conversation yet).

**Comment by paales** [comment]
We've build a module to address this issue some years ago: https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer

- Changes the query used: https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer/blob/master/src/Plugin/ChangelogPlugin.php#L39-L44
- Has better crash recovery if the indexers fails on a certain batch, we assume previous batches are fine: https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer/blob/master/src/Plugin/ViewPlugin.php#L60

And a few other fixes I don't specifically remember.

**Comment by hostep** [comment]
Confirmed over here as well. Seeing this problem on a shop we recently upgraded from Magento 2.3.5 to 2.4.5-p2

After some quick local testing, the module https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer seems to improve it a lot (with a backlog of 3861 items in the `catalog_product_attribute` indexer, default Magento 2.4.5 loops **203** times over the collection using a batch size of 1000, which is insane. The ReachDigital_BetterIndexer only needs **4** loops, which totally makes sense.

Haven't tested out the mentioned PR yet, will try to do that as well.

**Comment by hostep** [comment]
I've tested [the PR](https://github.com/magento/magento2/pull/36155) as well and it seems to be even more efficient than the ReachDigital_BetterIndexer module.

With the same dataset, the PR only takes **4** loops and fills up the loops with 1000 items always (except for the last loop which just contains the remainder). The ReachDigital_BetterIndexer module however didn't fill up every loop with 1000 items, only the first and third loop contained 1000 items in my case, the others were ~650, ~750 and ~450 items big.


So both are good solutions at first sight, we'll probably try out the PR on a real project in production soon-ish, as the ReachDigital_BetterIndexer contains more than just this bugfix and I'm a bit hesitant of adding more changes then only a fix for this problem without investigating them in more detail.


Thanks to all the people who commented here with your findings and proposed solutions, it really helped.
Let's hope Adobe will also start taking this issue a bit more seriously, because it's a really big problem in my opinion that causes a lot of unnecessary resource usage on servers for bigger Magento projects when using the partial reindexers.

**Comment by paales** [comment]
> The ReachDigital_BetterIndexer module however didn't fill up every loop with 1000 items, only the first and third loop contained 1000 items in my case, the others were ~650, ~750 and ~450 items big.

Not sure if there is any perf difference measurable the batch size is choosen arbitrarily at 1000, so no guaranteers that more will be faster.

> as the ReachDigital_BetterIndexer contains more than just this bugfix

Yeah it contains another important step, it can be that memory becomes an issue when running very large batches or faulty indexes. 

By default it will only save the indexer staat _after_ everything is indexed:
https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/lib/internal/Magento/Framework/Mview/View.php#L309

In de indexer it will save after every batch, which makes recovering a lot easier.
https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer/blob/master/src/Plugin/ViewPlugin.php#L167-L169

Also on crash it will not leave the indexer in a processing state:
https://github.com/ho-nl/magento2-ReachDigital_BetterIndexer/blob/master/src/Plugin/ViewPlugin.php#L197

**Comment by hostep** [comment]
Found a new quality patch that Adobe released yesterday, which is most likely also related to this issue: https://github.com/magento/quality-patches/compare/6ad18382c1...763cd71220#diff-e7a95f179c6968989ee14e35fa76700d6674f0fb45b9aac96b844df61f7c14f9

Description:
> ACSD-51114 : Fixes the issue where a random product disappears from large catalogs when asynchronous indexing is enabled. Improves performance of asynchronous reindexing for large catalogs.

---

Also another potential fix was mentioned in https://github.com/magento/magento2/issues/37367

---

Haven't tried any of these yet, but they sound promising as well...

**Comment by ihor-sviziev** [comment]
And one more related issue just reported https://github.com/magento/magento2/issues/37398

**Comment by dudzio12** [comment]
I've prepared PR implementing order by clause in select query just yesterday https://github.com/magento/magento2/pull/37399. It's funny that official patch ACSD-50814 was prepared also yesterday, since that bug is oldie one.

I've pointed in #37398 that implementing order by won't fix it completely. When there is more than one change resulting in inserting to _cl table, there will be that many records with this id. This is a situation where MySQL query with distinct feature is useless, because of the usage of limit in it. The best approach to fix it IMHO is to implement batching with PHP, make values unique and then chunk and reindex it.

**Comment by dudzio12** [comment]
Also in many indexers there are implemented batching features, so maybe we could completely remove batching on view update action, and just move responsibility to indexers itself.

ex.
`vendor/magento/module-catalog/Model/Indexer/Category/Flat/Action/Rows.php:147`,
`vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Rows.php:100`

But there are some indexes where batching is not implemented:
`vendor/magento/module-catalog-inventory/Model/Indexer/Stock/Action/Rows.php:24`

**Comment by hostep** [comment]
Some status update in case people are interested.

We've been recently trying out [quality patch from Adobe](https://github.com/magento/quality-patches/blob/763cd7122089ec46d6a18d51511ad4962f1339b3/patches/os/ACSD-51114_2.4.3-p2.patch) which is also supposed to fix this issue.
We've been running it on a Magento 2.4.3-p3 shop for a couple of days, at first look it seems to work fine, but it's now been 2 days in a row that we've noticed that the index cronjob is just sitting there for hours not doing anything while waiting on the following SQL query to finish:

```sql
DELETE cl_duplicate
FROM catalog_product_attribute_cl AS cl_duplicate,
     catalog_product_attribute_cl AS cl_origin
WHERE cl_duplicate.version_id < cl_origin.version_id
  AND cl_duplicate.version_id > 93264739
  AND cl_duplicate.version_id <= 93496654
  AND cl_origin.version_id > 93264739
  AND cl_origin.version_id <= 93496654
  AND cl_duplicate.entity_id <=> cl_origin.entity_id
```

When I attempt to run this query on my local machine with dbdump from the server, it's also just running for many many minutes without seemingly to progress in any way (I've stopped it after more than 10 minutes).
The `catalog_product_attribute_cl` table in our case contains 226253 rows and an EXPLAIN query returns this:
```
mysql> EXPLAIN DELETE cl_duplicate FROM catalog_product_attribute_cl AS cl_duplicate, catalog_product_attribute_cl AS cl_origin WHERE cl_duplicate.version_id < cl_origin.version_id AND cl_duplicate.version_id > 93264739 AND cl_duplicate.version_id <= 93496654 AND cl_origin.version_id > 93264739 AND cl_origin.version_id <= 93496654 AND cl_duplicate.entity_id <=> cl_origin.entity_id;
+----+-------------+--------------+------------+-------+---------------+---------+---------+------+--------+----------+-------------+
| id | select_type | table        | partitions | type  | possible_keys | key     | key_len | ref  | rows   | filtered | Extra       |
+----+-------------+--------------+------------+-------+---------------+---------+---------+------+--------+----------+-------------+
|  1 | DELETE      | cl_duplicate | NULL       | range | PRIMARY       | PRIMARY | 4       | NULL | 113126 |   100.00 | Using where |
|  1 | SIMPLE      | cl_origin    | NULL       | range | PRIMARY       | PRIMARY | 4       | NULL | 113126 |    10.00 | Using where |
+----+-------------+--------------+------------+-------+---------------+---------+---------+------+--------+----------+-------------+
2 rows in set (0.00 sec)
```

(I'm not a db expert, so I can't really interpret these results, so just providing for extra info)

So it sounds like that quality patch is not fully fixing this issue here, I think we'll switch back to the [community PR](https://github.com/magento/magento2/pull/36155), that one seems to work better (which we have been running for some weeks on another project, using Magento 2.4.5-p2)

**Comment by ilnytskyi** [comment]
Whoever interested to try adaptation of @dooblem approach use these patches:

```
Index: vendor/magento/framework/Mview/View.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vendor/magento/framework/Mview/View.php b/vendor/magento/framework/Mview/View.php
--- a/vendor/magento/framework/Mview/View.php
+++ b/vendor/magento/framework/Mview/View.php	(date 1701886312378)
@@ -314,7 +314,21 @@
                 break;
             }
             $vsFrom += $batchSize;
-            $action->execute($ids);
+
+            //TASKID: port of https://github.com/magento/magento2/issues/30012#issuecomment-1829364973 approach
+            $nextVersionId = intval(max(array_column($ids, 'max_version_id') ?: [0]));
+            if ($nextVersionId > 0) {
+                $vsFrom = $nextVersionId;
+            }
+
+            $isMultidimensional = isset($ids[0]) && is_array($ids[0]);
+
+            if ($isMultidimensional) {
+                $action->execute(array_map('intval', array_column($ids, 'entity_id')));
+            } else {
+                $action->execute($ids);
+            }
+            $this->getState()->setVersionId($vsFrom)->save();
         }
     }
```

and walker using max version id and having

```
Index: vendor/magento/framework/Mview/View/ChangeLogBatchWalker.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vendor/magento/framework/Mview/View/ChangeLogBatchWalker.php b/vendor/magento/framework/Mview/View/ChangeLogBatchWalker.php
--- a/vendor/magento/framework/Mview/View/ChangeLogBatchWalker.php
+++ b/vendor/magento/framework/Mview/View/ChangeLogBatchWalker.php	(date 1701886718709)
@@ -40,8 +40,8 @@
         if (!$connection->isTableExists($changelogTableName)) {
             throw new ChangelogTableNotExistsException(new Phrase("Table %1 does not exist", [$changelogTableName]));
         }
-
-        $select = $connection->select()->distinct(true)
+        //TASKID: adaptation of https://github.com/magento/magento2/issues/30012#issuecomment-1829364973 approach
+        $select = $connection->select()
             ->where(
                 'version_id > ?',
                 $fromVersionId
@@ -50,10 +50,16 @@
                 'version_id <= ?',
                 $toVersion
             )
+            //TASKID: adaptation of https://github.com/magento/magento2/issues/30012#issuecomment-1829364973 approach
+            ->having('max_version_id > ?', $fromVersionId)
+            ->having('max_version_id <= ?', $toVersion)
             ->group([$changelog->getColumnName()])
+            //TASKID: adaptation of https://github.com/magento/magento2/issues/30012#issuecomment-1829364973 approach
+            ->order('max_version_id ASC')
             ->limit($batchSize);
+        //TASKID: adaptation of https://github.com/magento/magento2/issues/30012#issuecomment-1829364973 approach
+        $select->from($changelogTableName, [$changelog->getColumnName(), 'max_version_id' => new \Zend_Db_Expr('MAX(version_id)')]);

-        $select->from($changelogTableName, [$changelog->getColumnName()]);
-        return $connection->fetchCol($select);
+        return $connection->fetchAll($select);
     }
 }
```

To me works like a rocket. We have about 5K products but update them frequently so changelogs are huge more that 200K, so now instead of walking CL by 1000 any partial reindex can process the whole catalog just in 5 iteration. Additionally after every iteration the Mview version is saved to DB.

**Comment by leonhelmus** [comment]
@hostep Did the pull request:
https://github.com/magento/magento2/pull/36155

Resolve most of the index issues? 
For one of my projects i see that the "Product price" Index gets stuck. 
M2 commerce: 2.4.6-p8
PHP: 8.1
mysql 8.0
customer groups: 4
products ~ 60k
Stores: 12

I have seen in my project where the "product price" gets stuck and it looks like when looking in mysql.
There is a insert query in catalog_product_index_price that continues to try to fill the table, but it looks like a never ending loop.
Once i invalidate the index "product_price" it gets unstuck and working again. This seems quite similar to what here is discussed. So i'm thinking of applying https://github.com/magento/magento2/pull/36155 as well, since it was also merged in the core.

Is there already a official patch released for this pr?

**Comment by hostep** [comment]
@leonhelmus: we indeed turned #36155 in a custom patch for our shops running on Magento 2.4.3 - 2.4.6.
No idea if it will fix your particular issue, but it should at least have some positive impact.

**Comment by engcom-Hotel** [comment]
Hello @ilnytskyi,

I suggest you to create a new issue explaining the same and with the steps to reproduce. We will proceed with that issue further.

Thanks
